File: backend/app/schemas/api_endpoint.py
"""
API 接口管理 Pydantic Schemas

定义接口管理的请求/响应数据结构,与 SQLAlchemy Model 解耦。
所有 JSON 字段(如 headers_json)在 schema 层用 dict/list 表达,序列化由 repository 处理。
"""
from typing import Any, Dict, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field, field_validator


# ============================================================
# 基础字段类型
# ============================================================

class HeaderItem(BaseModel):
    """请求头项"""
    key: str = Field(..., description="Header 名,如 Content-Type")
    value: str = Field(default="", description="Header 值")
    desc: str = Field(default="", description="说明")


class ParamItem(BaseModel):
    """参数项(Query/Path)"""
    name: str = Field(..., description="参数名")
    location: str = Field(default="query", description="参数位置: query/path/header/cookie")
    type: str = Field(default="string", description="参数类型: string/integer/boolean/number/array/object")
    required: bool = Field(default=False, description="是否必填")
    default: Optional[str] = Field(default=None, description="默认值")
    desc: str = Field(default="", description="说明")
    example: Optional[str] = Field(default=None, description="示例值")


class BodySpec(BaseModel):
    """请求体规范"""
    content_type: str = Field(default="application/json", description="Content-Type: application/json/form-data/x-www-form-urlencoded/raw")
    raw: Optional[str] = Field(default=None, description="原始文本(当 content_type=raw 时)")
    json_schema: Optional[Dict[str, Any]] = Field(default=None, description="JSON Schema 规范")
    form_fields: Optional[List[Dict[str, Any]]] = Field(default=None, description="表单字段列表")
    example: Optional[Any] = Field(default=None, description="示例")


class ResponseSpec(BaseModel):
    """响应规范"""
    status_code: int = Field(default=200, description="HTTP 状态码")
    headers: Optional[List[HeaderItem]] = Field(default=None, description="响应头")
    body_schema: Optional[Dict[str, Any]] = Field(default=None, description="响应体 JSON Schema")
    example: Optional[Any] = Field(default=None, description="响应示例")
    desc: str = Field(default="", description="说明")


class AuthSpec(BaseModel):
    """认证规范"""
    type: str = Field(default="none", description="认证类型: none/basic/bearer/api_key/oauth2/custom")
    details: Optional[Dict[str, Any]] = Field(default=None, description="认证详情(敏感字段用占位符)")


# ============================================================
# 创建/更新请求
# ============================================================

class EndpointCreate(BaseModel):
    """创建接口请求"""
    name: str = Field(..., min_length=1, max_length=200, description="接口名称")
    method: str = Field(..., description="HTTP 方法: GET/POST/PUT/DELETE/PATCH/HEAD/OPTIONS")
    path: str = Field(..., min_length=1, max_length=500, description="接口路径")
    summary: Optional[str] = Field(default=None, max_length=500, description="摘要")
    description: Optional[str] = Field(default=None, description="详细说明")
    tags: Optional[List[str]] = Field(default=None, description="标签列表")
    module: Optional[str] = Field(default=None, max_length=100, description="所属模块")
    status: str = Field(default="draft", description="状态: draft/active/deprecated/archived")
    source: str = Field(default="manual", description="来源: manual/swagger/postman/har/import/ai")
    headers: Optional[List[HeaderItem]] = Field(default=None, description="请求头")
    params: Optional[List[ParamItem]] = Field(default=None, description="Query/Path 参数")
    body: Optional[BodySpec] = Field(default=None, description="请求体")
    response: Optional[ResponseSpec] = Field(default=None, description="响应")
    auth: Optional[AuthSpec] = Field(default=None, description="认证")

    @field_validator("method")
    @classmethod
    def validate_method(cls, v: str) -> str:
        v = v.upper()
        allowed = {"GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"}
        if v not in allowed:
            raise ValueError(f"method 必须是 {allowed} 之一")
        return v

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        allowed = {"draft", "active", "deprecated", "archived"}
        if v not in allowed:
            raise ValueError(f"status 必须是 {allowed} 之一")
        return v


class EndpointUpdate(BaseModel):
    """更新接口请求(所有字段可选)"""
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    method: Optional[str] = Field(default=None)
    path: Optional[str] = Field(default=None, min_length=1, max_length=500)
    summary: Optional[str] = Field(default=None, max_length=500)
    description: Optional[str] = Field(default=None)
    tags: Optional[List[str]] = Field(default=None)
    module: Optional[str] = Field(default=None, max_length=100)
    status: Optional[str] = Field(default=None)
    headers: Optional[List[HeaderItem]] = Field(default=None)
    params: Optional[List[ParamItem]] = Field(default=None)
    body: Optional[BodySpec] = Field(default=None)
    response: Optional[ResponseSpec] = Field(default=None)
    auth: Optional[AuthSpec] = Field(default=None)

    @field_validator("method")
    @classmethod
    def validate_method(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        v = v.upper()
        allowed = {"GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"}
        if v not in allowed:
            raise ValueError(f"method 必须是 {allowed} 之一")
        return v

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        allowed = {"draft", "active", "deprecated", "archived"}
        if v not in allowed:
            raise ValueError(f"status 必须是 {allowed} 之一")
        return v


class EndpointQueryParams(BaseModel):
    """接口列表查询参数"""
    keyword: Optional[str] = Field(default=None, description="关键词,匹配 name/path/summary")
    method: Optional[str] = Field(default=None, description="HTTP 方法过滤")
    module: Optional[str] = Field(default=None, description="模块过滤")
    status: Optional[str] = Field(default=None, description="状态过滤")
    tags: Optional[List[str]] = Field(default=None, description="标签过滤(任一匹配)")
    page: int = Field(default=1, ge=1, description="页码,从 1 开始")
    page_size: int = Field(default=20, ge=1, le=100, description="每页条数")


class PublishRequest(BaseModel):
    """发布接口(生成版本快照)"""
    change_log: str = Field(default="", max_length=500, description="变更说明")


# ============================================================
# 响应
# ============================================================

class EndpointResponse(BaseModel):
    """接口详情响应"""
    id: int
    name: str
    method: str
    path: str
    summary: Optional[str] = None
    description: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    module: Optional[str] = None
    status: str
    source: str
    headers: List[HeaderItem] = Field(default_factory=list)
    params: List[ParamItem] = Field(default_factory=list)
    body: Optional[BodySpec] = None
    response: Optional[ResponseSpec] = None
    auth: AuthSpec = Field(default_factory=lambda: AuthSpec(type="none"))
    version: int
    user_id: Optional[int] = None
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class EndpointListItem(BaseModel):
    """接口列表项(精简字段)"""
    id: int
    name: str
    method: str
    path: str
    summary: Optional[str] = None
    module: Optional[str] = None
    status: str
    source: str
    version: int
    tags: List[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class EndpointListResponse(BaseModel):
    """接口列表响应"""
    total: int
    page: int
    page_size: int
    items: List[EndpointListItem]


class EndpointVersionItem(BaseModel):
    """接口版本项"""
    id: int
    endpoint_id: int
    version: int
    change_log: Optional[str] = None
    is_current: bool
    created_by: Optional[int] = None
    created_at: datetime


class EndpointVersionDetail(BaseModel):
    """接口版本详情(含快照)"""
    id: int
    endpoint_id: int
    version: int
    snapshot: Dict[str, Any]
    change_log: Optional[str] = None
    is_current: bool
    created_by: Optional[int] = None
    created_at: datetime


class EndpointStatsResponse(BaseModel):
    """接口统计响应"""
    total: int = Field(description="接口总数")
    by_status: Dict[str, int] = Field(default_factory=dict, description="按状态统计")
    by_method: Dict[str, int] = Field(default_factory=dict, description="按方法统计")
    by_module: Dict[str, int] = Field(default_factory=dict, description="按模块统计")


class BatchImportResult(BaseModel):
    """批量导入结果"""
    total: int = Field(description="解析到的接口数")
    imported: int = Field(default=0, description="成功导入数")
    skipped: int = Field(default=0, description="跳过数(method+path 重复)")
    failed: int = Field(default=0, description="失败数")
    errors: List[str] = Field(default_factory=list, description="错误详情")
    imported_ids: List[int] = Field(default_factory=list, description="导入的接口 ID 列表")

File: frontend/src/services/apiEndpoint.ts
/**
 * API 接口管理 - 前端服务层
 *
 * 对应后端路由前缀: /api/endpoints
 * 由于 request.ts 中 baseURL 已配置为 '/api', 本文件路径前缀为 '/endpoints'
 *
 * 覆盖接口:
 *   1. CRUD       - createEndpoint / listEndpoints / getEndpoint / updateEndpoint / deleteEndpoint
 *   2. 状态管理    - publishEndpoint / changeStatus
 *   3. 版本管理    - listVersions / getVersion / rollbackToVersion
 *   4. 统计       - getStats
 *   5. 批量导入    - batchImport
 */
import request from './request';

// ============================================================
// 类型定义 (与后端 schemas/api_endpoint.py 对齐)
// ============================================================

export type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS';
export type EndpointStatus = 'draft' | 'active' | 'deprecated' | 'archived';
export type EndpointSource = 'manual' | 'swagger' | 'postman' | 'har' | 'import' | 'ai';
export type AuthType = 'none' | 'basic' | 'bearer' | 'api_key' | 'oauth2' | 'custom';

export interface HeaderItem {
  key: string;
  value: string;
  desc?: string;
}

export interface ParamItem {
  name: string;
  location?: 'query' | 'path' | 'header' | 'cookie';
  type?: 'string' | 'integer' | 'boolean' | 'number' | 'array' | 'object';
  required?: boolean;
  default?: string | null;
  desc?: string;
  example?: string | null;
}

export interface BodySpec {
  content_type?: 'application/json' | 'form-data' | 'x-www-form-urlencoded' | 'raw';
  raw?: string | null;
  json_schema?: Record<string, unknown> | null;
  form_fields?: Array<Record<string, unknown>> | null;
  example?: unknown;
}

export interface ResponseSpec {
  status_code?: number;
  headers?: HeaderItem[] | null;
  body_schema?: Record<string, unknown> | null;
  example?: unknown;
  desc?: string;
}

export interface AuthSpec {
  type?: AuthType;
  details?: Record<string, unknown> | null;
}

// ============================================================
// 子表数据结构 (后端重构后: headers / body / parameters 拆分为独立子表)
// ============================================================

export interface ApiHeaderItem {
  id?: number;
  api_id?: number;
  key: string;
  value: string | null;
  required: boolean;
  sort_order: number;
}

export interface ApiBodyItem {
  id?: number;
  api_id?: number;
  body_type: string;
  schema_json: string | object | null;
  example_json: string | object | null;
  raw_text: string | null;
}

export interface ApiParameterItem {
  id?: number;
  api_id?: number;
  location: string;
  name: string;
  type: string;
  required: boolean;
  default_value: string | null;
  description: string | null;
  example: string | null;
  sort_order: number;
}

// 接口详情
export interface ApiEndpoint {
  id: number;
  name: string;
  method: HTTPMethod;
  path: string;
  summary?: string | null;
  description?: string | null;
  tags: string[];
  module?: string | null;
  status: EndpointStatus;
  source: EndpointSource;
  headers: HeaderItem[];
  params: ParamItem[];
  body?: BodySpec | null;
  response?: ResponseSpec | null;
  auth: AuthSpec;
  // 子表数据 (后端重构后优先使用,旧 JSON 字段保留以兼容)
  api_headers?: ApiHeaderItem[];
  api_body?: ApiBodyItem | null;
  api_parameters?: ApiParameterItem[];
  version: number;
  user_id?: number | null;
  created_by?: number | null;
  created_at: string;
  updated_at: string;
}

// 列表项 (精简字段)
export interface ApiEndpointListItem {
  id: number;
  name: string;
  method: HTTPMethod;
  path: string;
  summary?: string | null;
  module?: string | null;
  status: EndpointStatus;
  source: EndpointSource;
  version: number;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ApiEndpointListResponse {
  total: number;
  page: number;
  page_size: number;
  items: ApiEndpointListItem[];
}

export interface ApiEndpointStats {
  total: number;
  by_status: Record<string, number>;
  by_method: Record<string, number>;
  by_module: Record<string, number>;
}

export interface ApiEndpointVersionItem {
  id: number;
  endpoint_id: number;
  version: number;
  change_log?: string | null;
  is_current: boolean;
  created_by?: number | null;
  created_at: string;
}

export interface ApiEndpointVersionDetail extends ApiEndpointVersionItem {
  snapshot: Record<string, unknown>;
}

export interface BatchImportResult {
  total: number;
  imported: number;
  skipped: number;
  failed: number;
  errors: string[];
  imported_ids: number[];
}

// 创建/更新请求
export interface EndpointCreateInput {
  name: string;
  method: HTTPMethod | string;
  path: string;
  summary?: string;
  description?: string;
  tags?: string[];
  module?: string;
  status?: EndpointStatus;
  source?: EndpointSource;
  headers?: HeaderItem[];
  params?: ParamItem[];
  body?: BodySpec;
  response?: ResponseSpec;
  auth?: AuthSpec;
}

export type EndpointUpdateInput = Partial<EndpointCreateInput>;

export interface EndpointListParams {
  keyword?: string;
  method?: string;
  module?: string;
  status?: string;
  tags?: string[];
  page?: number;
  page_size?: number;
}

// ============================================================
// API 调用
// ============================================================

// ---- CRUD ----

export async function createEndpoint(data: EndpointCreateInput): Promise<ApiEndpoint> {
  return request.post('/endpoints', data);
}

export async function listEndpoints(params: EndpointListParams = {}): Promise<ApiEndpointListResponse> {
  const query: Record<string, unknown> = {
    page: params.page ?? 1,
    page_size: params.page_size ?? 20,
  };
  if (params.keyword) query.keyword = params.keyword;
  if (params.method) query.method = params.method;
  if (params.module) query.module = params.module;
  if (params.status) query.status = params.status;
  if (params.tags && params.tags.length) query.tags = params.tags.join(',');
  return request.get('/endpoints/list', { params: query });
}

export async function getEndpoint(id: number): Promise<ApiEndpoint> {
  return request.get(`/endpoints/${id}`);
}

export async function updateEndpoint(id: number, data: EndpointUpdateInput): Promise<ApiEndpoint> {
  return request.put(`/endpoints/${id}`, data);
}

export async function deleteEndpoint(id: number): Promise<void> {
  return request.delete(`/endpoints/${id}`);
}

// ---- 状态管理 ----

export async function publishEndpoint(id: number, changeLog: string = ''): Promise<ApiEndpoint> {
  return request.post(`/endpoints/${id}/publish`, { change_log: changeLog });
}

export async function changeEndpointStatus(id: number, newStatus: EndpointStatus): Promise<ApiEndpoint> {
  return request.put(`/endpoints/${id}/status`, undefined, {
    params: { new_status: newStatus },
  });
}

// ---- 版本管理 ----

export async function listEndpointVersions(id: number): Promise<ApiEndpointVersionItem[]> {
  return request.get(`/endpoints/${id}/versions`);
}

export async function getEndpointVersion(id: number, version: number): Promise<ApiEndpointVersionDetail> {
  return request.get(`/endpoints/${id}/versions/${version}`);
}

export async function rollbackEndpoint(id: number, version: number): Promise<ApiEndpoint> {
  return request.post(`/endpoints/${id}/rollback/${version}`);
}

// ---- 统计 ----

export async function getEndpointStats(): Promise<ApiEndpointStats> {
  return request.get('/endpoints/stats');
}

// ---- 批量导入 ----

export async function batchImportEndpoints(
  endpoints: EndpointCreateInput[],
  source: EndpointSource = 'import',
): Promise<BatchImportResult> {
  return request.post('/endpoints/batch-import', endpoints, { params: { source } });
}
